### PR TITLE
Fix projection progress report.

### LINKED ISSUE #2820: Projections reports their % done incorrectly in the UI
**Describe the bug**
Projections will report its percentage % completion of events processed wrong and if you write an event, then the projections will trigger and go to 100%, but without it, the projections don't do that recalculation.

**To Reproduce**
1. Start a EventStoreDB 20.10.2 node 
2. Enable all projections to see incorrect % done status

**Expected behavior**
 Process events and calculate percentage done correctly.  

**Actual behavior**

 Projection stats UI showing arbitrary numbers % done 

**Config/Logs/Screenshots**

image
**EventStore details**
 - EventStore server version: 
 
 Event Store 20.10.2.0

 - Operating system: Mac/ Linux /windows

 - EventStore client version (if applicable):
 
 N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by YoEight**
Fixed: Fix projection progress report.

The issue was about how `ReaderSubscriptionBase` reported progression. `ReaderSubscriptionBase` uses a time window to decide if we should report progress downstream. If that time windows was missed, progression was not reported.

The new solution reports progression asynchronously, every 500ms, to match previous time window.

Fixes #2820

### CODE REVIEW

File: src/EventStore.Projections.Core/Messages/ReaderSubscriptionMessage.cs [hunk 1]
```diff
@@ -265,5 +265,15 @@ public class Faulted : SubscriptionMessage {
...
 		}
+
+		public class ReportProgress : SubscriptionMessage {
+			private static readonly int TypeId = System.Threading.Interlocked.Increment(ref NextMsgId);
+
+			public override int MsgTypeId {
+				get { return TypeId; }
+			}
+
+			public ReportProgress() : base(Guid.NewGuid(), null, null) { }
```
Comment (thefringeninja) [inline src/EventStore.Projections.Core/Messages/ReaderSubscriptionMessage.cs:276]: IMO we should follow the convention established by the other implementations of `SubscriptionMessage`. `correlationId` and an optional `source` should come from the constructor.

File: src/EventStore.Projections.Core/Services/Processing/EventReaderCoreService.cs [hunk 2]
```diff
@@ -243,7 +245,17 @@ public class EventReaderCoreService :
...
-		
+		public void Handle(ReaderSubscriptionMessage.ReportProgress message) {
+			if (_stopped)
+				return;
+
+			foreach (var subscription in _subscriptions.Values) {
+				subscription.Handle(message);
+			}
+
+			_publisher.Publish(TimerMessage.Schedule.Create(TimeSpan.FromMilliseconds(500), new SendToThisEnvelope(this), new ReaderSubscriptionMessage.ReportProgress()));
```
Comment (thefringeninja) [inline src/EventStore.Projections.Core/Services/Processing/EventReaderCoreService.cs:256]: See above comment. These messages should share the same correlation id

Comment (thefringeninja) [reply src/EventStore.Projections.Core/Services/Processing/EventReaderCoreService.cs:256]: Also, should we be creating a new envelope every time?

**Review by thefringeninja — CHANGES_REQUESTED** [review]

File: src/EventStore.Projections.Core/Messages/ReaderSubscriptionMessage.cs (hunk 1 above)
Comment (YoEight) [reply src/EventStore.Projections.Core/Messages/ReaderSubscriptionMessage.cs]: Done

File: src/EventStore.Projections.Core/Services/Processing/EventReaderCoreService.cs (hunk 2 above)
Comment (YoEight) [reply src/EventStore.Projections.Core/Services/Processing/EventReaderCoreService.cs]: Done

**Review by hayley-jean — CHANGES_REQUESTED** [review]
This is reporting the progress better, but it still doesn't dependably reach 100%

The projection only updates its progress when it handles an event that passes its filter. If there are a large number of events that don't pass the projection's filters at the end of the stream/$all then the progress will get stuck again.

Probably the easiest way to test this is to write a bunch of system events at the end of the log. The projections will report the progress correctly while the events are being written, but if you restart the node afterwards, the progress will be incorrect again.

Setting the progress to 100% once the reader receives the `EventReaderIdle` message might be the simplest way to solve this, since the reader should be caught up when it goes idle. However, this needs to be tested in case the readers also enter idle if it's paused to wait for a checkpoint.

**Review by hayley-jean — APPROVED** [review]

**Review by thefringeninja — APPROVED** [review]

**Pull request merged**